{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "943f580a",
   "metadata": {},
   "source": [
    "# Tutorial Pleiotropy Correction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d9ab843",
   "metadata": {},
   "source": [
    "# Step 1: Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "33b966ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/friva/miniconda3/envs/pyviper_dev/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<module 'pyviper' from '/mnt/beegfs/home/friva/pyviper/pyviper/__init__.py'>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# --- Core scientific stack ---\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import norm\n",
    "import matplotlib.pyplot as plt\n",
    "# --- For single-cell data if your workflow uses AnnData / Scanpy ---\n",
    "import scanpy as sc\n",
    "# --- Your local pyVIPER package (the one you edited) ---\n",
    "import pyviper\n",
    "import importlib\n",
    "import pyviper._viper as _v  # this gives you direct access to the edited _viper.py\n",
    "# --- Reload to make sure Python uses the updated code ---\n",
    "importlib.reload(_v)\n",
    "importlib.reload(pyviper)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8629f35",
   "metadata": {},
   "source": [
    "# Step 2: Import Gene expression matrix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "53c40265",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1106 × 19938"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set the HPC file path\n",
    "gene_expr_path = \"/home/friva/TCGA_DEG/DEG-BRCA.tsv\"\n",
    "# Read the data\n",
    "gene_expr_signature = pd.read_csv(gene_expr_path, sep=\"\\t\", index_col=0)\n",
    "# Convert to AnnData object\n",
    "gene_expr_signature = sc.AnnData(gene_expr_signature)\n",
    "# Inspect\n",
    "gene_expr_signature"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e125b07",
   "metadata": {},
   "source": [
    "# Step 3: Import ARACNe Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b76ab50d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>target</th>\n",
       "      <th>mor</th>\n",
       "      <th>likelihood</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCG1</td>\n",
       "      <td>PLEKHF2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.986344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCG1</td>\n",
       "      <td>WWP1</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.981491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCG1</td>\n",
       "      <td>SERPINB5</td>\n",
       "      <td>-0.999987</td>\n",
       "      <td>0.977465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCG1</td>\n",
       "      <td>HRCT1</td>\n",
       "      <td>-0.999980</td>\n",
       "      <td>0.970370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCG1</td>\n",
       "      <td>PLAGL1</td>\n",
       "      <td>-0.999981</td>\n",
       "      <td>0.970235</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  regulator    target       mor  likelihood\n",
       "0     ABCG1   PLEKHF2  1.000000    0.986344\n",
       "1     ABCG1      WWP1  0.999999    0.981491\n",
       "2     ABCG1  SERPINB5 -0.999987    0.977465\n",
       "3     ABCG1     HRCT1 -0.999980    0.970370\n",
       "4     ABCG1    PLAGL1 -0.999981    0.970235"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "network_path = \"/home/friva/TCGA_networks/net-prunned-BRCA_test.tsv\"\n",
    "\n",
    "network = pd.read_csv(network_path, delimiter=\"\\t\")\n",
    "\n",
    "# Drop the first column if it's just an index\n",
    "network = network.drop(network.columns[0], axis=1)\n",
    "\n",
    "network_interactome = pyviper.Interactome('TCGA_BRCA', network)  # convert to class Interactome\n",
    "network_interactome.size()  # show number of regulons in the Interactome\n",
    "network_interactome.net_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fc454e5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removed 0 targets.\n",
      "Removed 15308 targets.\n",
      "Removed 412 regulators.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/beegfs/home/friva/pyviper/pyviper/interactome.py:638: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  pruned_df = sorted_df.groupby('regulator', group_keys=False).apply(lambda x: x.iloc[:max_targets])\n"
     ]
    }
   ],
   "source": [
    "# keep targets present in expression\n",
    "network_interactome.filter_targets(gene_expr_signature.var_names)\n",
    "# prune to exactly 50 targets per TF (drop TFs with <50)\n",
    "network_pruned = network_interactome.copy()\n",
    "network_pruned.prune(max_targets=50, eliminate=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1662e180",
   "metadata": {},
   "source": [
    "# Step 4: Run PyViper with \"pleiotropy = False\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dc72f716",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing the association scores\n",
      "Computing regulons enrichment with aREA\n"
     ]
    }
   ],
   "source": [
    "BRCA_NES = pyviper.viper(gex_data=gene_expr_signature, # gene expression signature\n",
    "                             interactome=network_pruned, # gene regulatory network\n",
    "                             enrichment = \"area\",\n",
    "                             min_targets=0,\n",
    "                             eset_filter=True,\n",
    "                             pleiotropy = False,\n",
    "                             njobs=10,\n",
    "                             verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c298e8e",
   "metadata": {},
   "source": [
    "# Step 5: Run PyViper with \"pleiotropy = True\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c2c84a86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing the association scores\n",
      "Computing regulons enrichment with aREA\n",
      "[pleiotropy] hook reached\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Initializing regulons: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10400/10400 [00:00<00:00, 31571.96it/s]\n",
      "Filling regulons: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10400/10400 [00:00<00:00, 29072.93it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[pleiotropy] ss shape: (19938, 1106)\n",
      "[pleiotropy] NES0 shape (TFs×samples): (208, 1106)\n",
      "[pleiotropy] workers: 10\n",
      "[pleiotropy] updated 726/1106 samples (median TFs updated: 1)\n"
     ]
    }
   ],
   "source": [
    "BRCA_NES_Corrected = pyviper.viper(gex_data=gene_expr_signature, # gene expression signature\n",
    "                             interactome=network_pruned, # gene regulatory network\n",
    "                             enrichment = \"area\",\n",
    "                             min_targets=0,\n",
    "                             eset_filter=True,\n",
    "                             pleiotropy = True,\n",
    "                             njobs=10,\n",
    "                             verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "71ef178b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 3.31299488,  0.13898613,  1.41328845, ...,  4.27563766,\n",
       "         4.27255937,  1.81609743],\n",
       "       [ 4.40232445,  0.17952647,  1.66693234, ...,  2.76371511,\n",
       "        -1.09122135, -0.25404518],\n",
       "       [-1.00436247,  2.15137492,  3.10341038, ..., -0.20926719,\n",
       "        -0.04146679, -0.08405256],\n",
       "       ...,\n",
       "       [ 0.85092515, -0.61553694,  1.4120394 , ..., -0.49405199,\n",
       "        -2.23862397, -2.20046406],\n",
       "       [ 4.4411637 , -0.4607928 ,  0.82933544, ...,  2.90994808,\n",
       "         8.02378342,  4.86682743],\n",
       "       [-5.08276292, -0.07092113, -3.33750214, ..., -8.08333506,\n",
       "        -8.76794786, -8.41765852]], shape=(1106, 208))"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BRCA_NES.X\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "617cafde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 3.31299488,  0.13898613,  1.41328845, ...,  4.27563766,\n",
       "         4.27255937,  1.81609743],\n",
       "       [ 4.40232445,  0.17952647,  1.66693234, ...,  2.76371511,\n",
       "        -1.09122135, -0.25404518],\n",
       "       [-1.00436247,  2.15137492,  3.10341038, ..., -0.20926719,\n",
       "        -0.04146679, -0.08405256],\n",
       "       ...,\n",
       "       [ 0.85092515, -0.61553694,  1.4120394 , ..., -0.49405199,\n",
       "        -2.23862397, -2.20046406],\n",
       "       [ 4.4411637 , -0.4607928 ,  0.82933544, ...,  2.90994808,\n",
       "         8.02378342,  4.86682743],\n",
       "       [-5.08276292, -0.07092113, -3.33750214, ..., -8.08333506,\n",
       "        -8.76794786, -8.41765852]], shape=(1106, 208))"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BRCA_NES_Corrected.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "db99f11d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Convert AnnData matrix (samples × TFs) to DataFrame\n",
    "BRCA_NES_df = pd.DataFrame(\n",
    "    BRCA_NES.X,\n",
    "    index=BRCA_NES.obs_names,\n",
    "    columns=BRCA_NES.var_names\n",
    ")\n",
    "\n",
    "# Transpose if you prefer TFs as rows and samples as columns\n",
    "BRCA_NES_df = BRCA_NES_df.T\n",
    "\n",
    "# Save to CSV\n",
    "BRCA_NES_df.to_csv(\"NES_pyVIPER_precorrection_BRCA_github.csv\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "23da6f02",
   "metadata": {},
   "outputs": [],
   "source": [
    "BRCA_NES_Corrected_df = pd.DataFrame(\n",
    "    BRCA_NES_Corrected.X,\n",
    "    index=BRCA_NES_Corrected.obs_names,\n",
    "    columns=BRCA_NES_Corrected.var_names\n",
    ")\n",
    "\n",
    "# Transpose if you prefer TFs as rows and samples as columns\n",
    "BRCA_NES_Corrected_df = BRCA_NES_Corrected_df.T\n",
    "\n",
    "# Save to CSV\n",
    "BRCA_NES_Corrected_df.to_csv(\"NES_pyVIPER_postcorrection_BRCA_github.csv\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34126f42",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
